sw/device/sca/ecc256_keygen_serial.c

```diff
@@ -0,0 +1,302 @@
+};
+
+OTBN_DECLARE_APP_SYMBOLS(p256_key_from_seed_sca);
+
+OTBN_DECLARE_SYMBOL_ADDR(p256_key_from_seed_sca, mode);
+OTBN_DECLARE_SYMBOL_ADDR(p256_key_from_seed_sca, seed0);
+OTBN_DECLARE_SYMBOL_ADDR(p256_key_from_seed_sca, seed1);
+OTBN_DECLARE_SYMBOL_ADDR(p256_key_from_seed_sca, d0);
+OTBN_DECLARE_SYMBOL_ADDR(p256_key_from_seed_sca, d1);
+OTBN_DECLARE_SYMBOL_ADDR(p256_key_from_seed_sca, x);
+OTBN_DECLARE_SYMBOL_ADDR(p256_key_from_seed_sca, y);
+
+static const otbn_app_t kOtbnAppP256KeyFromSeed =
+    OTBN_APP_T_INIT(p256_key_from_seed_sca);
+
+static const otbn_addr_t kOtbnVarMode =
+    OTBN_ADDR_T_INIT(p256_key_from_seed_sca, mode);
+static const otbn_addr_t kOtbnVarSeed0 =
+    OTBN_ADDR_T_INIT(p256_key_from_seed_sca, seed0);
+static const otbn_addr_t kOtbnVarSeed1 =
+    OTBN_ADDR_T_INIT(p256_key_from_seed_sca, seed1);
+static const otbn_addr_t kOtbnVarD0 =
+    OTBN_ADDR_T_INIT(p256_key_from_seed_sca, d0);
+static const otbn_addr_t kOtbnVarD1 =
+    OTBN_ADDR_T_INIT(p256_key_from_seed_sca, d1);
+static const otbn_addr_t kOtbnVarX =
+    OTBN_ADDR_T_INIT(p256_key_from_seed_sca, x);
+static const otbn_addr_t kOtbnVarY =
+    OTBN_ADDR_T_INIT(p256_key_from_seed_sca, y);
+
+/**
+ * Seed value.
+ *
+ * The default value corresponds to the test data in
+ *   sw/otbn/crypto/test/p256_key_from_seed_test.s
+ *
+ * This default value can be overwritten via the simpleserial command `s`
+ * (see ecc256_set_seed)
+ */
+uint32_t ecc256_seed[kEcc256SeedNumWords] = {
+    0x016064e9, 0x11e3f4d6, 0xac3a6fa7, 0xaba11a1b, 0x8f9271d1,
+    0x22b79d5f, 0x1176f31d, 0xb5ac3a51, 0x99a082d7, 0x484eb366,
+};
+
+/**
+ * Simple serial 'x' (set seed) command handler.
+ *
+ * The key must be `kEcc256SeedNumBytes` bytes long.
+ *
+ * @param seed Value for seed share.
+ * @param seed_len Length of seed share.
+ */
+static void ecc256_set_seed(const uint8_t *seed, size_t seed_len) {
+  SS_CHECK(seed_len == kEcc256SeedNumBytes);
+  memcpy(ecc256_seed, seed, seed_len);
+}
+
+/**
+ * Runs the OTBN key generation program.
+ *
+ * The seed shares must be `kEcc256SeedNumWords` words long.
+ *
+ * @param[in] mode  Mode parameter (private key only or full keypair).
+ * @param[in] seed  Seed for key generation.
+ * @param[in] mask  Mask for seed.
+ */
+static void p256_run_keygen(uint32_t mode, const uint32_t *seed,
+                            const uint32_t *mask) {
+  // Write mode.
+  SS_CHECK_STATUS_OK(otbn_dmem_write(/*num_words=*/1, &mode, kOtbnVarMode));
+
+  // Compute first share of seed (seed ^ mask).
+  uint32_t seed0[kEcc256SeedNumWords];
+  for (size_t i = 0; i < kEcc256SeedNumWords; i++) {
+    seed0[i] = seed[i] ^ mask[i];
+  }
+
+  // Write seed shares.
+  SS_CHECK_STATUS_OK(
+      otbn_dmem_write(kEcc256SeedNumWords, seed0, kOtbnVarSeed0));
+  SS_CHECK_STATUS_OK(otbn_dmem_write(kEcc256SeedNumWords, mask, kOtbnVarSeed1));
+
+  // Execute program.
+  sca_set_trigger_high();
+  SS_CHECK_STATUS_OK(otbn_execute());
+  SS_CHECK_STATUS_OK(otbn_busy_wait_for_done());
```

Comment:
nit: In the past I have found Ibex busy polling to be very noisy for SCA measurements. `sca_call_and_sleep` is the latest incarnation of our call and sleep function that avoids busy polling. For OTBN we could even have a wake up on OTBN done which could further improve the interface. 

Since in this case OTBN itself could be a source of noise, I am not sure how relevant this is, but overall I have found reducing noise to be overall beneficial for white box testing.

Reply:
Hmm, good idea. I tried to make this change but it looks complex enough that I think it's best to do it in a follow-up, since folks are building on top of this code already and I don't want to cause rebase problems. In particular, I'm not sure how best to calculate the overhead time it takes OTBN to wipe its internal state and send the `done` interrupt after an `ecall` instruction. Created #17338 to track.